packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
  * @param compatBehavior - Typed redundantly with CompatModeBehavior to ensure handling is added when updating that type
  */
 function compatBehaviorAllowsMessageType(
-	_unknownContainerRuntimeMessageType: never,
+	_unknownContainerRuntimeMessageType: UnknownContainerMessageType,
 	compatBehavior: "Ignore" | "FailToProcess" | undefined,
 ): boolean {
 	// undefined defaults to same behavior as "FailToProcess"
 	return compatBehavior === "Ignore";
 }
 
 /**
- * The unpacked runtime message / details to be handled or dispatched by the ContainerRuntime
+ * The unpacked runtime message / details to be handled or dispatched by the ContainerRuntime.
+ * Message type are differentiated via a `type` string and contain different contents depending on their type.
  *
  * IMPORTANT: when creating one to be serialized, set the properties in the order they appear here.
  * This way stringified values can be compared.
  */
-export interface ContainerRuntimeMessage {
+export interface ContainerRuntimeMessage<TType, TContents> {
 	/** Type of the op, within the ContainerRuntime's domain */
-	type: ContainerMessageType;
+	type: TType;
 	/** Domain-specific contents, interpreted according to the type */
-	contents: any;
+	contents: TContents;
 	/** Info describing how to handle this op in case the type is unrecognized (default: fail to process) */
 	compatDetails?: IContainerRuntimeMessageCompatDetails;
 }
 
+export type ContainerRuntimeDataStoreOpMessage = ContainerRuntimeMessage<
+	ContainerMessageType.FluidDataStoreOp,
+	IEnvelope
+>;
+export type InboundContainerRuntimeAttachMessage = ContainerRuntimeMessage<
+	ContainerMessageType.Attach,
+	InboundAttachMessage
+>;
+export type OutboundContainerRuntimeAttachMessage = ContainerRuntimeMessage<
+	ContainerMessageType.Attach,
+	IAttachMessage
+>;
+export type ContainerRuntimeChunkedOpMessage = ContainerRuntimeMessage<
+	ContainerMessageType.ChunkedOp,
+	IChunkedOp
+>;
+export type ContainerRuntimeBlobAttachMessage = ContainerRuntimeMessage<
+	ContainerMessageType.BlobAttach,
+	undefined
+>;
+export type ContainerRuntimeRejoinMessage = ContainerRuntimeMessage<
+	ContainerMessageType.Rejoin,
+	undefined
+>;
+export type ContainerRuntimeAliasMessage = ContainerRuntimeMessage<
+	ContainerMessageType.Alias,
+	IDataStoreAliasMessage
+>;
+export type LocalContainerRuntimeIdAllocationMessage = ContainerRuntimeMessage<
+	ContainerMessageType.IdAllocation,
+	IdCreationRangeWithStashedState
+>;
+export type InTransitContainerRuntimeIdAllocationMessage = ContainerRuntimeMessage<
```

Comment:
I was thinking about this, maybe we can drop the `InTransit` from the name - the type over the wire can be the "canonical" type, and then the `Local`-named version has extra.  Or if you like it leave it, that phrase "in transit" was my idea but I never really like it :P

Reply:
FWIW I am good with dropping InTransit- and just keeping Local-.